### PR TITLE
Add <hollow-cylinder> as recognised shape in XML definitions

### LINKED ISSUE #16132: Add <hollow-cylinder> tag to list of known shapes
In the XML definition of shape an hollow-cylinder, of finite length, can be created from the difference of two cylinders. It would be more convenient to have a specific tag to create such a shape.

This will be more frequently used for defining cans in sample environments.

The suggested definition follows that of a [cylinder](http://docs.mantidproject.org/nightly/concepts/HowToDefineGeometricShape.html#howtodefinegeometricshape#cylinder):

```xml
<hollow-cylinder id="an-1">
  <centre-of-bottom-base x="0.0" y="0.0" z="0.0" />
  <axis x="0.0" y="1.0" z="0" />
  <inner-radius val="0.006"/>
  <outer-radius val="0.0065"/>
  <height val="1" />
</hollow-cylinder>
```

### PR DESCRIPTION
**Pull request by martyngigg**
The `ShapeFactory` has been extended to understand the concept of a `hollow-cylinder`. See the referenced issue for more detail.

**To test:**

Review the code and check the example in the unit tests makes sense.

As a test for visualization I modified the MARI instrument definition locally to use a hollow cylinder instead of a cylinder for the fermi-chopper. The instrument view then shows the expected shape:

hollowcylinderinsrtumentview

Try something similar to convince yourself it works as expected.

Fixes #16132

[Release notes updated](/mantidproject/mantid/commit/c7770f9c3b534949b08ce01c9129bbab32fc9425)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/src/Objects/ShapeFactory.cpp [hunk 1]
```diff
@@ -504,6 +507,73 @@ std::string ShapeFactory::parseSegmentedCylinder(
...
+
+  // add bottom plane
+  auto pPlaneBottom = boost::make_shared<Plane>();
+  pPlaneBottom->setPlane(centreOfBottomBase, normVec);
+  prim[l_id] = pPlaneBottom;
+  retAlgebraMatch << "" << l_id << ")";
+  l_id++;
+
+  return retAlgebraMatch.str();
+}
```
Comment (mantid-roman) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: Are we checking for `innerRadius < outerRadius`? Or the largest becomes the outer?

Comment (martyngigg) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: We're not checking at the moment. None of the other definitions do either, which is equally bad. I was a little loathed to restructure the code too much at this stage as it's core stuff.

Do you think this is worth doing?

Comment (mantid-roman) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: Maybe in the new code only?

Comment (martyngigg) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: @mantid-roman I've added some checks for the new type and tests that they fail as expected.

Comment (mantid-roman) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: Zero height behaves OK in the instrument view: no visible shape.

Zero or negative radii give "Unknown exception".

Comment (martyngigg) [inline Framework/Geometry/src/Objects/ShapeFactory.cpp]: Nice catch. I'll update the error checks.

**Comment by martyngigg** [comment]
@mantid-roman I've added a few more checks and I don't see any exceptions in the instrument view any longer.

**Comment by mantid-roman** [comment]
:shipit:

**Comment by FedeMPouzols** [comment]
Nice checks and tests against bad inputs!

**Pull request merged**